tests/components/alarm_control_panel/test_manual_mqtt.py

```diff
@@ -1395,53 +1395,60 @@ def test_state_changes_are_published_to_mqtt(self):
 
         # Component should send disarmed alarm state on startup
         self.hass.block_till_done()
-        self.assertEqual(('alarm/state', STATE_ALARM_DISARMED, 0, True),
-                         self.mock_publish.mock_calls[-2][1])
+        self.mock_publish.async_publish.assert_called_once_with(
```

Comment:
These too are dangerous. Rename it to `assert_called_once_with_bla_bla` and the test will still pass.

Reply:
(means the old code was 👍)

Also, when doing a refactor, don't try to change the tests too. That makes it harder to validate that your code did not change anything.

Reply:
I understand your concern. There were however reasons for this change. Let me explain.

> These too are dangerous. Rename it to `assert_called_once_with_bla_bla` and the test will still pass.

This is actually not entirely true. Are these asserts dangerous in principle (when used incorrectly)? Yes, very much so (*). Are they dangerous here? Not so much. For example, renaming the method to `assert_called_once_with_bla_bla` here will correctly raise an `AttributeError`.

<img width="824" alt="screen shot 2018-02-10 at 11 26 48" src="https://user-images.githubusercontent.com/6833237/36061401-ebdb8020-0e59-11e8-93d4-52847dd32d57.png">

This is because I'm **wrapping** the MQTT client object in a MagicMock and setting [**`spec_set`**](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock) as well. By doing this, "attempting to set or get an attribute on the mock that isn’t on the object passed as *spec_set* will raise an `AttributeError`."

(* Side Note: Starting with Python 3.5, `MagicMock`s have an argument called `unsafe` that when set to `False` (the default) causes an `AttributeError` to be raised when attempting a non-existant call on a attribute with the name starting with `assert` or `assret`)

> (means the old code was 👍)

When updating the base tests to use the actual MQTT client code and not the dispatcher, *something* changed and all `self.mock_publish.mock_calls[-2][1]` asserts were failing. Why did all of them suddenly fail? I didn't know at the time and it seemed very weird to me at the time since the *publish* code didn't really change. Why was the second last mock_call always accessed by the test code with `[-2]`? - I had no idea at all.

Of course, especially after the previous comment about code reviews being complicated by style fixes, I wanted to change as little as possible for the test code - after all, it would also mean quite a bit of work for me. Anyway, so I spend quite a while trying to debug the problem and (only partially successfully) spinning up a debugger (I'm no Python developer, to say the least) - without finding a solution using mock_calls. And after a while I found the "potential" fix of just using `assert_called_once_with`, which to me also seemed to have the added benefit of being much easier to read - especially so since I personally had struggled quite a bit understanding what `[-2]` was trying to indicate.

Frankly, I have to say that at the time, I didn't know about the potential pitfalls of using `assert_called_*` and setting `spec_set` was mostly *accidental* and of the motto: "it can't hurt".

I know now how I could go about "fixing" the tests without changing too much code - simply changing `-2` to `-1` in `self.mock_publish.mock_calls[-2][1]` *seems* to work. I do, however, *not* know *why* the `-2` was there in the first place - that would definitely be good to know before going about fixing it that way.

> That makes it harder to validate that your code did not change anything.

I completely understand. I hope you understand that at the time and after unsuccessfully trying to fix the tests, it seemed like the best option to me.

Now, I can, if you want to, change the tests to use `[-1]` everywhere or if you know why `-2` was used everywhere, I could also try to fix it that way.

Reply:
Ok, I didn't know about the `spec_set`, in that case it's better as it's more clear what is happening.

The reason things like -2 etc were used was because some other code was also calling something on the MQTT mock.

BTW, instead of going all fancy with the debugger, just `print(self.mock_publish.mock_calls)` will show a lot :)

Reply:
Yeah I also stumbled upon that by accident, but it's quite useful. And not using `print(...)` at first was kind of dumb of me but hey you learn something new every day 😊